virtual_ecosystem/models/animal/animal_cohorts.py

```diff
@@ -1418,3 +1436,64 @@ def is_migration_season(self) -> bool:
         """
 
         return random.random() <= self.constants.seasonal_migration_probability
+
+    def match_vertical(self, resource_occupancy: VerticalOccupancy) -> bool:
+        """Check whether cohort vertical occupancy overlaps with a resource or prey.
+
+        This method determines whether the vertical occupancy of the consumer cohort
+        overlaps with the vertical occupancy of a resource (pool or cohort). Animals
+        can only forage resources that share at least one overlapping vertical space.
+
+        Args:
+            resource_occupancy: The vertical occupancy trait of the potential resource
+                or prey.
+
+        Returns:
+            True if the vertical occupancy overlaps; False otherwise.
+
+        """
+        occupancy_pairs = {
+            VerticalOccupancy.SOIL: {
+                VerticalOccupancy.SOIL,
+                VerticalOccupancy.SOIL_GROUND,
+                VerticalOccupancy.SOIL_GROUND_CANOPY,
+            },
+            VerticalOccupancy.GROUND: {
+                VerticalOccupancy.GROUND,
+                VerticalOccupancy.SOIL_GROUND,
+                VerticalOccupancy.GROUND_CANOPY,
+                VerticalOccupancy.SOIL_GROUND_CANOPY,
+            },
+            VerticalOccupancy.CANOPY: {
+                VerticalOccupancy.CANOPY,
+                VerticalOccupancy.GROUND_CANOPY,
+                VerticalOccupancy.SOIL_GROUND_CANOPY,
+            },
+            VerticalOccupancy.SOIL_GROUND: {
+                VerticalOccupancy.SOIL,
+                VerticalOccupancy.GROUND,
+                VerticalOccupancy.SOIL_GROUND,
+                VerticalOccupancy.GROUND_CANOPY,
+                VerticalOccupancy.SOIL_GROUND_CANOPY,
+            },
+            VerticalOccupancy.GROUND_CANOPY: {
+                VerticalOccupancy.GROUND,
+                VerticalOccupancy.CANOPY,
+                VerticalOccupancy.SOIL_GROUND,
+                VerticalOccupancy.GROUND_CANOPY,
+                VerticalOccupancy.SOIL_GROUND_CANOPY,
+            },
+            VerticalOccupancy.SOIL_GROUND_CANOPY: {
+                VerticalOccupancy.SOIL,
+                VerticalOccupancy.GROUND,
+                VerticalOccupancy.CANOPY,
+                VerticalOccupancy.SOIL_GROUND,
+                VerticalOccupancy.GROUND_CANOPY,
+                VerticalOccupancy.SOIL_GROUND_CANOPY,
+            },
+        }
+
+        return (
+            resource_occupancy
+            in occupancy_pairs[self.functional_group.vertical_occupancy]
+        )
```

Comment:
If you implement the `VerticalOccupancy` the way I suggest, these whole thing becomes:

```suggestion
        return bool(
            resource_occupancy & self.functional_group.vertical_occupancy]
        )
```

All the logic to define the validity is taken care for by the `Flag` enum. This might also come handy in other cases.